rsocket-core/src/main/java/io/rsocket/RSocketClient.java

```diff
@@ -35,44 +32,96 @@
 
 import java.nio.channels.ClosedChannelException;
 import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.concurrent.atomic.AtomicInteger;
 import java.util.function.Consumer;
 import java.util.function.Function;
 import java.util.function.Supplier;
 
 /**
  * Client Side of a RSocket socket. Sends {@link Frame}s
- * to a {@link ServerRSocket}
+ * to a {@link RSocketServer}
  */
-public class ClientRSocket implements RSocket {
+class RSocketClient implements RSocket {
 
     private static final ClosedChannelException CLOSED_CHANNEL_EXCEPTION = new ClosedChannelException();
 
     private final DuplexConnection connection;
     private final Consumer<Throwable> errorConsumer;
     private final StreamIdSupplier streamIdSupplier;
-    private final KeepAliveProvider keepAliveProvider;
     private final MonoProcessor<Void> started;
     private final IntObjectHashMap<LimitableRequestPublisher> senders;
     private final IntObjectHashMap<Subscriber<Payload>> receivers;
+    private final AtomicInteger missedAckCounter;
 
     private Disposable keepAliveSendSub;
-    private volatile Consumer<Lease> leaseConsumer; // Provided on start()
 
-    public ClientRSocket(DuplexConnection connection, Consumer<Throwable> errorConsumer,
-                                StreamIdSupplier streamIdSupplier, KeepAliveProvider keepAliveProvider) {
+    private volatile long timeLastTickSentMs;
+
+    RSocketClient(DuplexConnection connection,
+                         Consumer<Throwable> errorConsumer,
+                         StreamIdSupplier streamIdSupplier) {
+        this(connection, errorConsumer, streamIdSupplier, Duration.ZERO, Duration.ZERO, 0);
+    }
+
+    RSocketClient(DuplexConnection connection,
+                         Consumer<Throwable> errorConsumer,
+                         StreamIdSupplier streamIdSupplier,
+                         Duration tickPeriod,
+                         Duration ackTimeout,
+                         int missedAcks) {
         this.connection = connection;
-        this.errorConsumer = new KnownErrorFilter(errorConsumer);
+        this.errorConsumer = errorConsumer;
         this.streamIdSupplier = streamIdSupplier;
-        this.keepAliveProvider = keepAliveProvider;
         this.started = MonoProcessor.create();
         this.senders = new IntObjectHashMap<>(256, 0.9f);
         this.receivers = new IntObjectHashMap<>(256, 0.9f);
+        this.missedAckCounter = new AtomicInteger();
+
+        if (tickPeriod != Duration.ZERO) {
```

Comment:
A straight object reference comparison seems unsafe here.  

 * This is a <a href="{@docRoot}/java/lang/doc-files/ValueBased.html">value-based</a>
 * class; use of identity-sensitive operations (including reference equality
 * ({@code ==}), identity hash code, or synchronization) on instances of
 * {@code Duration} may have unpredictable results and should be avoided.
 * The {@code equals} method should be used for comparisons.